sharkfuser/tests/test_matmul_node.cpp

```diff
@@ -0,0 +1,515 @@
+    attr.setC(std::make_shared<TensorAttr>());
+
+    MatmulNode node(std::move(attr), ctx);
+    FUSILLI_REQUIRE_OK(node.preValidateNode());
+    FUSILLI_REQUIRE_OK(node.inferPropertiesNode());
+    FUSILLI_REQUIRE_OK(node.postValidateNode());
+
+    auto cT = node.matmulAttr.getC();
+    // Output should have the broadcasted batch dimension with contiguous stride
+    REQUIRE(cT->getDim() == std::vector<int64_t>{batch, m, n});
+    REQUIRE(cT->getStride() == std::vector<int64_t>{m * n, n, 1});
+  }
+
+  SECTION("A has batch dimension of 1, B has larger batch dimension") {
+    int64_t batch = 8;
+
+    attr.setA(std::make_shared<TensorAttr>(
+        TensorAttr().setDim({1, m, k}).setStride({m * k, k, 1})));
+    attr.setB(std::make_shared<TensorAttr>(
+        TensorAttr().setDim({batch, k, n}).setStride({k * n, n, 1})));
+    // C is under-specified (dim/stride missing).
+    attr.setC(std::make_shared<TensorAttr>());
+
+    MatmulNode node(std::move(attr), ctx);
+    FUSILLI_REQUIRE_OK(node.preValidateNode());
+    FUSILLI_REQUIRE_OK(node.inferPropertiesNode());
+    FUSILLI_REQUIRE_OK(node.postValidateNode());
+
+    auto cT = node.matmulAttr.getC();
+    // Output should have the broadcasted batch dimension (max of 1 and batch)
+    REQUIRE(cT->getDim() == std::vector<int64_t>{batch, m, n});
+    REQUIRE(cT->getStride() == std::vector<int64_t>{m * n, n, 1});
+  }
+
+  SECTION("Multiple batch dimensions with broadcasting") {
+    int64_t batch1 = 4, batch2 = 8;
+
+    attr.setA(std::make_shared<TensorAttr>(
+        TensorAttr()
+            .setDim({1, batch2, m, k})
+            .setStride({batch2 * m * k, m * k, k, 1})));
+    attr.setB(
+        std::make_shared<TensorAttr>(TensorAttr()
+                                         .setDim({batch1, 1, k, n})
+                                         .setStride({k * n, k * n, n, 1})));
+    // C is under-specified (dim/stride missing).
+    attr.setC(std::make_shared<TensorAttr>());
+
+    MatmulNode node(std::move(attr), ctx);
+    FUSILLI_REQUIRE_OK(node.preValidateNode());
+    FUSILLI_REQUIRE_OK(node.inferPropertiesNode());
+    FUSILLI_REQUIRE_OK(node.postValidateNode());
+
+    auto cT = node.matmulAttr.getC();
+    // Output should have broadcasted batch dimensions: max(1, batch1),
+    // max(batch2, 1)
+    REQUIRE(cT->getDim() == std::vector<int64_t>{batch1, batch2, m, n});
+    REQUIRE(cT->getStride() ==
+            std::vector<int64_t>{batch2 * m * n, m * n, n, 1});
+  }
+}
+
+TEST_CASE("MatmulNode preValidate accepts arbitrary input strides",
+          "[matmul_node]") {
+  Context ctx;
+  MatmulAttr attr;
+
+  int64_t m = 16, k = 32, n = 64;
+
+  auto aT = std::make_shared<TensorAttr>(
+      TensorAttr().setDim({m, k}).setStride({k, 1}).setName("A_contiguous"));
+
+  auto bT = std::make_shared<TensorAttr>(
+      TensorAttr()
+          .setDim({k, n})
+          .setStride(
+              {k * n, k}) // Arbitrary layout (not contiguous or channels-last)
```

Comment:
What does it mean to have an arbitrary layout? Does this change how we emit the mlir asm / handle transposes.